integration/spark/src/main/scala/org/apache/spark/sql/CarbonCountStar.scala

```diff
@@ -108,4 +125,145 @@ case class CarbonCountStar(
+  // read again.
+  // Step 4. use multi-thread to prune segment and partition to get pruned index file list, which
+  // can prune most index files and reduce the files num.
+  // Step 5. read the count from pruned index file directly and cache it, get from cache if exist
+  // in the index_file <-> rowCount map.
+  private def getRowCountPurePartitionPrune: Long = {
+    var rowCount: Long = 0
+    val prunedPartitionPaths = new java.util.HashSet[String]()
+    // Get the current partitions from table.
+    val partitions = CarbonFilters.getPrunedPartitions(relation, predicates)
+    if (partitions != null) {
+      for (partition <- partitions) {
+        prunedPartitionPaths.add(partition.getLocation.toString)
+      }
+      val details = SegmentStatusManager.readLoadMetadata(carbonTable.getMetadataPath)
+      val validSegmentPaths = details.filter(segment =>
+        ((segment.getSegmentStatus == SegmentStatus.SUCCESS) ||
+          (segment.getSegmentStatus == SegmentStatus.LOAD_PARTIAL_SUCCESS))
+          && segment.getSegmentFile != null).map(segment => segment.getSegmentFile)
+      val tableSegmentIndexes = IndexStoreManager.getInstance().getAllSegmentIndexes(
+        carbonTable.getTableId)
+      if (!tableSegmentIndexes.isEmpty) {
+        // clear invalid cache
+        for (segmentFilePathInCache <- tableSegmentIndexes.keySet().asScala) {
+          if (!validSegmentPaths.contains(segmentFilePathInCache)) {
+            // means invalid cache
+            tableSegmentIndexes.remove(segmentFilePathInCache)
+          }
+        }
+      }
+      // init and put absent the valid cache
+      for (validSegmentPath <- validSegmentPaths) {
+        if (tableSegmentIndexes.get(validSegmentPath) == null) {
+          val segmentIndexMeta = new SegmentIndexMeta(validSegmentPath)
+          tableSegmentIndexes.put(validSegmentPath, segmentIndexMeta)
+        }
+      }
+
+      val numThreads = Math.min(Math.max(validSegmentPaths.length, 1), 4)
+      val executorService = Executors.newFixedThreadPool(numThreads)
+      // to get the index files of valid segments from cache or scan through each segment
+      val futures = new java.util.ArrayList[Future[Long]]
+      for (segmentPath <- validSegmentPaths) {
+        futures.add(executorService.submit(new Callable[Long] {
+          override def call(): Long = {
+            var rowCountCurrentSeg: Long = 0
+            var indexFilesCurrentSeg: java.util.List[CarbonFile] = null;
+            // tableSegmentIndexes already init.
+            if (tableSegmentIndexes.get(segmentPath) != null &&
+              tableSegmentIndexes.get(segmentPath).getSegmentIndexFiles.size() > 0) {
+              indexFilesCurrentSeg =
+                tableSegmentIndexes.get(segmentPath).getSegmentIndexFiles
+            } else {
+              // read from seg file
+              val sfs = new SegmentFileStore(carbonTable.getTablePath, segmentPath)
```

Comment:
This will give NullPointerException if the segmentFileName is null for a segment or  line number 191, `indexFilesCurrentSeg` will get null